### PR TITLE
[geo-awareness] [mock_uss] Initial Geo-Awareness Mock USS - Status and Geozone Source endpoint

### PR DESCRIPTION
**Pull request by barroco**
This PR adds an initial version of the geo-awareness mock_uss. It implements the /status and /geozone_source endpoints of the [geo-awareness automated testing interface](https://github.com/interuss/dss/blob/master/interfaces/automated-testing/geo-awareness/geo-awareness.yaml).
The geozone_source endpoint loads and parses ED269 formatted files and stores it "as is" in the in-memory database. This may be improved in a next PR addressing the checks endpoint.

To accomodate Common Information services, AUTH_SPEC and DSS_URL are now only required for collaborative services (RID and SCD).

### CODE REVIEW

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
+import json
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:1]: I think it's fine to leave this in for now, but I think this would be a strong candidate for moving to [uas_standards](https://github.com/interuss/uas_standards) when time allows.

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    purpose: Optional[Purpose]
+    intervalBefore: Optional[str]
+
+
+class VerticalReferenceType(str, Enum):
+    AGL = "AGL"
+    AMSL = "AMSL"
+
+
+# TODO: Discuss with @ben about Union[PolygonType, CircleType] which throws a not implemented exception
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:53]: Ug, implicit polymorphism is difficult :(  I don't think it's a particularly solvable problem in general (how would one tell if one was receiving an A or a B in general when the declared type was Union[A, B]?) and really it's probably not worth addressing in an automatic framework.  For the purpose of deserialization as part of a complex ImplicitDict type, the escape valve is to just declare the field a `dict`.  That way, there won't be any schema checking as part of ImplicitDict deserialization, but the data will still be available as a plain dict which can be checked manually.  ImplicitDict is just a fancy dict, so it never hurts to revert to a plain dict -- all that is lost is some automatic validation features and syntactic sugar.

TL;DR: Instead of `my_field: Union[PolygonType, CircleType]`, one possibility is to just do `my_field: dict`

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+#     radius: float # > 0
+
+# Start workaround
+class UniversalHorizontalProjectionType(str, Enum):
+    Circle = "Circle"
+    Polygon = "Polygon"
+
+
+class UniversalHorizontalProjection(ImplicitDict):
+    type: str
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:77]: This is a great workaround!  Probably better than my `dict` possibility above.  Should this be `type: UniversalHorizontalProjectionType`?  You can then add helper methods on `UniversalHorizontalProjection` that return `CircleType` and `PolygonType` if they're helpful.

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+#     center: List[float] # 2 items
+#     radius: float # > 0
+
+# Start workaround
+class UniversalHorizontalProjectionType(str, Enum):
+    Circle = "Circle"
+    Polygon = "Polygon"
+
+
+class UniversalHorizontalProjection(ImplicitDict):
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:76]: nit: I'd probably just name it `CircleOrPolygonType` (but that's a style preference)

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    THU = "THU"
+    FRI = "FRI"
+    SAT = "SAT"
+    SUN = "SUN"
+    ANY = "ANY"
+
+
+class DailyPeriod(ImplicitDict):
+    day: List[WeekDateType]  # min items: 1, max items: 7
+    startTime: str  # TODO: Convert to a Time representation
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:113]: `implicitdict.StringBasedDateTime` is a great solution here if the field is an RFC-3339 string.  It would look something like this (not checked for full accuracy):

```python
from implicitdict import StringBasedDateTime

class DailyPeriod(ImplicitDict):
    startTime: StringBasedDateTime

p = DailyPeriod(startTime=StringBasedDateTime(datetime.utcnow()))
p2 = ImplicitDict.parse(json.loads(json.dumps(p)), DailyPeriod)
assert p2.startTime.datetime == p.startTime.datetime
```

**Review by BenjaminPelletier — APPROVED** [review]
This looks great -- I've left comments, but approval indicates I'm confident you'll resolve them in a way that I would approve of.  Feel free to ask me to take a look at any changes you make if you'd like, but also feel free to simply merge when you've addressed or triaged the comments.

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py (hunk 1 above)
Comment (barroco) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:1]: Let's keep it here for few more PRs and I will move it there when mature.

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py (hunk 5 above)
Comment (barroco) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:113]: Unfortunately, startTime is only the time part of a ISO8601 datetime including the timezone (eg: `01:02:03.00Z`). Arrow do not support ISO8601 time only values.

Comment (barroco) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:113]: Correction, arrow is able to handle it. I will propose a new datastructure to the implicitdict repository. Let me know if you see a better approach.

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py (hunk 3 above)
Comment (barroco) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:77]: Done. Thanks. Will add helper in upcoming iterations if needed.

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py (hunk 4 above)
Comment (barroco) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:76]: Done

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    extendedProperties: Optional[Any]
+
+
+class ED269Schema(ImplicitDict):
+    title: Optional[str]
+    description: Optional[str]
+    features: List[UASZoneVersion]
+
+
+def parse(raw_data: Dict) -> ED269Schema:
```
[the remark this answers is not in the document]
Comment (barroco) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:150]: Done

File: monitoring/mock_uss/geoawareness/routes_geoawareness.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    get_geozone_source,
+    create_geozone_source,
+    delete_geozone_source,
+)
+from monitoring.monitorlib.geoawareness_automated_testing import api as geoawareness_api
+
+
+@webapp.route(
+    "/geoawareness/geozone_sources/<geozone_source_id>",
+    methods=["GET", "PUT", "DELETE"],
```
[the remark this answers is not in the document]
Comment (barroco) [reply monitoring/mock_uss/geoawareness/routes_geoawareness.py:17]: Done

File: monitoring/mock_uss/geoawareness/database.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        state: GeozoneSourceState,
+        message: Optional[str] = None,
+    ) -> SourceRecord:
+        with db as tx:
+            if id in tx.sources.keys():
+                raise ExistingRecordException()
+            tx.sources[id] = SourceRecord(
+                definition=definition, state=state, message=message
+            )
+        return db.value.sources[id]
```
[the remark this answers is not in the document]
Comment (barroco) [reply monitoring/mock_uss/geoawareness/database.py:49]: Done

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py (hunk 2 above)
Comment (barroco) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:53]: The fallback to simple dict is good to know, thanks :)

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    intervalBefore: Optional[str]
+
+
+class VerticalReferenceType(str, Enum):
+    AGL = "AGL"
+    AMSL = "AMSL"
+
+
+# TODO: Discuss with @ben about Union[PolygonType, CircleType] which throws a not implemented exception
+#  in implicitdict.
```
[the remark this answers is not in the document]
Comment (BenjaminPelletier) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py]: Oops, missed this comment the first time when I was reviewing but I think this is covered in another thread.

File: monitoring/mock_uss/geoawareness/routes_geoawareness.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+
+@webapp.route(
+    "/geoawareness/geozone_sources/<geozone_source_id>",
+    methods=["PUT"],
+)
+@requires_scope([geoawareness_api.SCOPE_GEOAWARENESS_TEST])
+def put_geozone_sources(geozone_source_id: str) -> Tuple[str, int]:
+    try:
+        json = flask.request.json
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/routes_geoawareness.py:31]: Should this be `json = flask.request.json()`?

**Review by BenjaminPelletier — APPROVED** [review]

File: monitoring/mock_uss/geoawareness/routes_geoawareness.py (hunk 10 above)
Comment (barroco) [reply monitoring/mock_uss/geoawareness/routes_geoawareness.py:31]: It is the [request](https://flask.palletsprojects.com/en/2.2.x/api/#flask.request) "[proxy](https://flask.palletsprojects.com/en/2.2.x/reqcontext/#notes-on-proxies)" of type Request with json as a [property](https://flask.palletsprojects.com/en/2.2.x/api/#flask.Request.json) which calls get_json() with default arguments.

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py (hunk 5 above)
Comment (barroco) [reply monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py]: Correction, arrow can parse the values but uses the minimum date `1900-01-01` as reference which can lead to a date out while manipulating timezones. Based on your input, I came up with a proposal. May I kindly ask you to check the last commit?

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 11]
```diff
@@ -87,10 +88,28 @@ class WeekDateType(str, Enum):
     ANY = "ANY"
 
 
+class ED269TimeType(str):
+    """String that allows values which describe a time in ED-269 flavour of ISO 86001 format.
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:92]: nit: 8601?

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 12]
```diff
@@ -87,10 +88,28 @@ class WeekDateType(str, Enum):
...
+
+    ED-269 standard specifies that a time instant type should be in the form of hh:mmS where S is
+    the timezone. However, examples are using the following format: 00:00:00.00Z
+    This class supports both formats.
+    """
+
+    def __new__(cls, value):
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:99]: I know `implicitdict` doesn't do this right now, but it would probably be useful to type-hint: `value: Union[str, time]` (or whatever the appropriate types are)

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 13]
```diff
@@ -87,10 +88,28 @@ class WeekDateType(str, Enum):
...
+        if isinstance(value, str):
+            t = arrow.get(value, ['HH:mm:ss.SZ', 'HH:mmZ']).timetz()
+        else:
+            t = value
+        str_value = str.__new__(cls, t.strftime('%H:%M:%S.%f')[:11] + t.strftime('%z').replace('+0000', 'Z'))
+        str_value.time = t
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:107]: Also something `implicitdict` doesn't do right now, but declaring this field in the class along with type-hinting would definitely help I think

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    TUE = "TUE"
+    WED = "WED"
+    THU = "THU"
+    FRI = "FRI"
+    SAT = "SAT"
+    SUN = "SUN"
+    ANY = "ANY"
+
+
+class ED269TimeType(str):
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:91]: Looks great to me :)

File: monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    """String that allows values which describe a time in ED-269 flavour of ISO 86001 format.
+
+    ED-269 standard specifies that a time instant type should be in the form of hh:mmS where S is
+    the timezone. However, examples are using the following format: 00:00:00.00Z
+    This class supports both formats.
```
Comment (BenjaminPelletier) [inline monitoring/mock_uss/geoawareness/parsers/ed269/__init__.py:96]: nit: "This class supports both formats as inputs and uses the hh:mmS as the output format."

**Review by BenjaminPelletier — APPROVED** [review]
